core/domain/suggestion_services.py

```diff
@@ -0,0 +1,367 @@
+
+
+def get_suggestions_by_author(author_id):
+    """Gets a list of suggestions by the given author.
+
+    Args:
+        author_id: str. The author of the suggestions.
+
+    Returns:
+        list(Suggestions): a list of suggestions by the given author.
+    """
+    return [
+        get_suggestion_from_model(s)
+        for s in suggestion_models.SuggestionModel.get_suggestions_by_author(
+            author_id)]
+
+
+def get_suggestions_reviewed_by(reviewer_id):
+    """Gets a list of suggestions that have been reviewed by the given user.
+
+    Args:
+        reviewer_id: str. The reviewer of the suggestion.
+
+    Returns:
+        list(Suggestions): a list of suggestions reviewed by the given user.
+    """
+    return [
+        get_suggestion_from_model(s)
+        for s in suggestion_models.SuggestionModel.get_suggestions_reviewed_by(
+            reviewer_id)]
+
+
+def get_suggestions_assigned_to_reviewer(assigned_reviewer_id):
+    """Gets a list of suggestions assigned to the given user for review.
+
+    Args:
+        assigned_reviewer_id: str. The reviewer assigned to review the
+                suggestion.
+
+    Returns:
+        list(Suggestions): a list of suggestions assigned to the given user
+            for review.
+    """
+    return [
+        get_suggestion_from_model(s)
+        for s in suggestion_models.SuggestionModel
+        .get_suggestions_assigned_to_reviewer(assigned_reviewer_id)]
+
+
+def get_suggestions_by_status(status):
+    """Gets a list of suggestions with the given status.
+
+    Args:
+        status: str. The status of the suggestion.
+
+    Returns:
+        list(Suggestions): a list of suggestions with the given status.
+    """
+    return [
+        get_suggestion_from_model(s)
+        for s in suggestion_models.SuggestionModel.get_suggestions_by_status(
+            status)]
+
+
+def get_suggestion_by_type(suggestion_type):
+    """Gets a list of suggestions with the given sub_type.
+
+    Args:
+        suggestion_type: str. The sub type of the suggestion.
+
+    Returns:
+        list(Suggestions): a list of suggestions of the given type.
+    """
+    return [
+        get_suggestion_from_model(s)
+        for s in suggestion_models.SuggestionModel.get_suggestions_by_type(
+            suggestion_type)]
+
+
+def get_suggestions_by_target_id(target_type, target_id):
+    """Gets a list of suggestions to the entity with the given ID.
+
+    Args:
+        target_type: str. The type of target entity the suggestion is linked to.
+        target_id: str. The ID of the target entity the suggestion is linked to.
+
+    Returns:
+        list(Suggestions): a list of suggestions linked to the entity.
+    """
+    return [
+        get_suggestion_from_model(s)
+        for s in suggestion_models.SuggestionModel.get_suggestions_by_target_id(
+            target_type, target_id)]
+
+
+def update_suggestion(suggestion):
+    """Updates the given sugesstion. The properties that can be edited are the
```

Comment:
"suggestion" is misspelled.

Also, the design here seems a bit weird. I think update_suggestion should be a private function that takes a Suggestion domain object that's already been prepared and just copies the fields blindly to the model. You can then have public functions that correspond to specific actions (like update_reviewer or mark_review_completed, etc.) and that call _update_suggestion in the background.

Reply:
I'll make this change to the design. Thanks!

Reply:
Actually I have one question here. Wouldn't this increase the number of write operations to the DB? If for each field update we call the _update_suggestion (which writes into the db), that would not be optimal right?

Reply:
I think it depends. It doesn't make sense to do this for multiple changes to the content, yep. But why would you call stuff like update_reviewer or mark_review_completed at the same time?

Basically, figure out what the "high-level actions" are -- mark review completed, start a suggestion, etc. and have a public function for each of them. That function does all the preparations to the domain object -- it doesn't need to limit its changes to just one field. Then it calls _update_suggestion to commit it.

Reply:
Ok this makes sense. I'll do this.

I was initially thinking to have one update_status and another update_reviewer and so on. Each of which calls _update_suggestion and then the accept_suggestion calls these functions. But this would have been inefficient. I'll go with your suggestion of having a function for each action.

Reply:
Done and made this a protected function.